File: vrp-platform/app/types/vrp.ts
export type Node = {
  id: number;
  lat: number;
  lng: number;
  x: number;
  y: number;
  demand: number;
};

export type VRPRequest = {
  nodes: Node[];
  depot_id: number;
  vehicle_capacity: number;
};

export type Route = {
  customers: number[];
  total_demand: number;
  total_distance: number;
  geometry: [number, number][];
  road_distance_km?: number;
  duration_s?: number;
};

export type MergeEvent = {
  i: number;
  j: number;
  routes: number[][];
  geometries?: [number, number][][];
};

export type SavingsRow = {
  i: number;
  j: number;
  saving: number;
};

export type VRPResponse = {
  routes: Route[];
  total_distance: number;
  total_road_distance_km?: number;
  total_duration_s?: number;
  num_vehicles?: number;
  computation_time_ms?: number;
  savings_table: SavingsRow[];
  steps: string[];
  merge_events?: MergeEvent[];
  edge_geometries?: Record<string, [number, number][]>;
};

export type VRPComparisonResponse = {
  clarke_wright: VRPResponse;
  cplex: VRPResponse;
};

File: backend/schemas/vrp.py
from pydantic import BaseModel
from typing import List, Optional, Any

class MergeEvent(BaseModel):
    i: int
    j: int
    routes: List[List[int]]


class Node(BaseModel):
    id: int
    lat: float
    lng: float
    x: float
    y: float
    demand: float = 0


class VRPRequest(BaseModel):
    nodes: List[Node]
    depot_id: int
    vehicle_capacity: float


class Route(BaseModel):
    customers: List[int]
    total_demand: float
    total_distance: float
    geometry: List[List[float]] = []
    road_distance_km: Optional[float] = None
    duration_s: Optional[float] = None

class VRPResponse(BaseModel):
    routes: List[Route]
    total_distance: float
    total_road_distance_km: Optional[float] = None
    total_duration_s: Optional[float] = None
    num_vehicles: int = 0
    computation_time_ms: Optional[float] = None
    savings_table: List[dict]
    steps: List[str]
    merge_events: Optional[List[MergeEvent]] = None
    # Optimized: key format "id1,id2" -> polyline
    edge_geometries: Optional[dict[str, List[List[float]]]] = None